### PR TITLE
Move Instance Construction To Type Classes

### PR DESCRIPTION
**Pull request by fppt**
- Move addEntity to EntityType
- Move addRelation to RelationType
- Move addRule to RuleType
- Move putResource to ResourceType

### CODE REVIEW

File: grakn-graph/src/main/java/ai/grakn/example/MovieGraphFactory.java [hunk 1]
```diff
@@ -156,109 +156,109 @@ private static void buildOntology() {
     }
 
     private static void buildInstances() throws ParseException {
-        godfather = graknGraph.addEntity(movie);
+        godfather = movie.addEntity();
```
Comment (aelred) [inline grakn-graph/src/main/java/ai/grakn/example/MovieGraphFactory.java]: Love this! Makes so much sense.

File: grakn-graql/src/main/java/ai/grakn/graql/internal/query/InsertQueryExecutor.java [hunk 2]
```diff
@@ -260,11 +261,11 @@ private Concept putConceptByType(Optional<String> id, VarAdmin var, Type type) {
      * @return an instance of the specified type, with the given ID if one was specified
      */
     private <T extends Type, S extends Instance> S addOrGetInstance(
-            Optional<String> id, T type, Function<String, S> getInstance, Function<T, S> addInstance
+            Optional<String> id, T type, Function<String, S> getInstance, Supplier<S> addInstance
```
Comment (aelred) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/query/InsertQueryExecutor.java]: You probably don't need to pass the type to this function anymore

**Comment by fppt** [comment]
Does than mean your rejection is due to requesting changes from yourself ? :)

**Comment by fppt** [comment]
@aelred Your changes seem to have made Jenkins unhappy. I am going to rebase and put this back up.

File: grakn-test/src/test/java/ai/grakn/test/graql/analytics/ClusteringTest.java [hunk 3]
```diff
@@ -345,55 +345,55 @@ private void addResourceRelations() throws GraknValidationException {
...
+                .putRolePlayer(resourceValue5, graph.getResourceType(resourceType5).putResource(-7L));
+        relationType5.addRelation()
                 .putRolePlayer(resourceOwner5, entity2)
-                .putRolePlayer(resourceValue5, graph.putResource(-7L, graph.getResourceType(resourceType5)));
-        graph.addRelation(relationType5)
+                .putRolePlayer(resourceValue5, graph.getResourceType(resourceType5).putResource(-7L));
+        relationType5.addRelation()
                 .putRolePlayer(resourceOwner5, entity4)
-                .putRolePlayer(resourceValue5, graph.putResource(-7L, graph.getResourceType(resourceType5)));
+                .putRolePlayer(resourceValue5, graph.getResourceType(resourceType6).putResource(-7L));
```
Comment (duckofyork) [inline grakn-test/src/test/java/ai/grakn/test/graql/analytics/ClusteringTest.java]: @fppt !!!!!!!!!!!

**Pull request merged**